### PR TITLE
fix for app catalog #2602

### LINKED ISSUE #2602: Issue with ALM api, Property 'getAppCatalog' does not exist
### Category
- [ ] Enhancement
- [ ] Bug
- [x] Question
- [ ] Documentation gap/issue

### Version

Please specify what version of the library you are using: [  3.2.1       ]

Please specify what version(s) of SharePoint you are targeting: [      Sharepoint Online   spfx 1.16]

*If you are not using the latest release, please update and see if the issue is resolved before submitting an issue.*

### Expected / Desired Behavior / Question
I am trying the ALM api examples, specifically the following:
```
import { spfi } from "@pnp/sp";
import "@pnp/sp/appcatalog";
import "@pnp/sp/webs";

const sp = spfi(...);

// you can get the tenant app catalog (or any app catalog) by using the getTenantAppCatalogWeb method
const appCatWeb = await sp.getTenantAppCatalogWeb()();
const appCatalog = await appCatWeb.getAppCatalog()();

// you can get the tenant app catalog (or any app catalog) by passing in a url
// get the tenant app catalog
const tenantCatalog = await sp.web.getAppCatalog("https://mytenant.sharepoint.com/sites/appcatalog")();

// get a different app catalog
const catalog = await sp.web.getAppCatalog("https://mytenant.sharepoint.com/sites/anothersite")();
```

I cant get past 'error TS2339: Property 'getAppCatalog' does not exisit on type 'IWeb'
Any help greatly appreciated!

I am new to 3.x I am wondering if this needs something special similar to logging ref : [sp-dev-fx-webparts](https://github.com/pnp/sp-dev-fx-webparts)/[samples](https://github.com/pnp/sp-dev-fx-webparts/tree/main/samples)/[react-pnp-js-sample](https://github.com/pnp/sp-dev-fx-webparts/tree/main/samples/react-pnp-js-sample)/[src](https://github.com/pnp/sp-dev-fx-webparts/tree/main/samples/react-pnp-js-sample/src)/[webparts](https://github.com/pnp/sp-dev-fx-webparts/tree/main/samples/react-pnp-js-sample/src/webparts)/[pnPjsExample](https://github.com/pnp/sp-dev-fx-webparts/tree/main/samples/react-pnp-js-sample/src/webparts/pnPjsExample)/pnpjsConfig.ts

### LINKED ISSUE #2615: Batching and Caching on Graph
### Category
- [ ] Enhancement
- [X] Bug
- [ ] Question
- [ ] Documentation gap/issue

### Version

Please specify what version of the library you are using: [3.13.0]
Please specify what version(s) of SharePoint you are targeting: [1.15.2]

### Expected / Desired Behavior / Question
The desired behavior is to get a cached value for a batched request with graph package.

### Observed Behavior
The batch is correctly executed the first time to get results from Microsoft Graph.
The cache correctly works (correct results stored into Local Storage)
The execution is stopped on `await execute()` on next execution of the batch.
No error is logged into the console.

On the screenshot below, the line 167 is never reached on the next executions of the batch (first time worked correctly):
image

### Steps to Reproduce

- I copy/paste the code from Batching Graph example from docs: https://pnp.github.io/pnpjs/concepts/batching/#graph-example
- I use the `SPFx` behavior because the code is run for a SharePoint WebPart
- I add the `Caching` behavior for the batch

```
import { graphfi, SPFx } from "@pnp/graph"; // this line has been updated to add SPFX
import { Caching } from '@pnp/queryable'; // this line has been added
import "@pnp/graph/users";
import "@pnp/graph/groups";
import "@pnp/graph/batching";

const graph = graphfi().using(SPFx(this.context)); // this line has been updated to add SPFX

const [batchedGraph, execute] = graph.batched();
batchedGraph.using(Caching()); // this line has been added

let res = [];

// Pushes the results of these calls to an array
// you need to use .then syntax here as otherwise the application will stop and await the result
batchedGraph.users().then(r => res.push(r));

// you need to use .then syntax here as otherwise the application will stop and await the result
// ODATA operations such as select, filter, and expand are supported as normal
batchedGraph.groups.select("Id")().then(r => res.push(r));

// Executes the batched calls
await execute();

// Results for all batched calls are available
for(let i=0; i<res.length; i++){
    // Do something with the results
    console.log(res[i]);
}
```

### PR DESCRIPTION
**Pull request by patrick-rodgers**
#### Category
- [X] Bug fix?
- [ ] New feature?
- [ ] New sample?
- [ ] Documentation update?

#### Related Issues

fixes #2602
fixes #2615 

#### What's in this Pull Request?

- adds a tenantAppcatalog property to the root sp object
- ensures that web.appcatalog points to the sitecollection appcatalog
- fixes a bug with batching and caching in graph where requests were not resolving correctly, the fix from #2052 applied to graph lib

This might be a breaking change, but it wasn't working correctly before - so should actually start doing what people expect.

### CODE REVIEW

File: docs/sp/alm.md [hunk 1]
```diff
@@ -18,48 +20,36 @@ import "@pnp/sp/webs";
...
 import { spfi } from "@pnp/sp";
 import { Web } from '@pnp/sp/webs';
-import { AppCatalog } from '@pnp/sp/appcatalog';
 
-const sp = spfi().using(...);
-const web = Web([sp.web, "https://mytenant.sharepoint.com/sites/apps"]);
-const catalog = await AppCatalog(web)();
+const sp = spfi(...);
+const web = Web([sp.web, "https://mytenant.sharepoint.com/sites/mysite"]);
+const catalog = await web.appcatalog();
```
Comment (bcameron1231) [inline docs/sp/alm.md:52]: Should this be clarified?

web.appcatalog() would return apps, web.catalog would return the catalog.

const sp = spfi(...);
const web = Web([sp.web, "https://mytenant.sharepoint.com/sites/mysite"]);
const catalog = web.appcatalog;
const apps = await catalog();

Comment (patrick-rodgers) [reply docs/sp/alm.md:52]: Not sure I follow. This is just assigning the `web.appcatalog` prop to `catalog`, then calling it. `catalog` can be `foo` But if its not clear we can update.

File: packages/sp/appcatalog/types.ts
[the remark this answers is not in the document]
Comment (patrick-rodgers) [reply packages/sp/appcatalog/types.ts:1]: Good catch, I missed this.

File: packages/sp/appcatalog/types.ts [hunk 2]
```diff
@@ -12,6 +12,20 @@ import { File, IFile } from "../files/types.js";
...
+    const paths = ["_api/web/tenantappcatalog/", "_api/web/sitecollectionappcatalog/"];
+
+    for (let i = 0; i < paths.length; i++) {
+        const index = base.indexOf(paths[i]);
+        if (index > -1) {
+            return combine(base.substring(index, index + paths[i].length), path);
+        }
+    }
+
+    return url;
```
[the remark this answers is not in the document]
Comment (patrick-rodgers) [reply packages/sp/appcatalog/types.ts]: Well shoot. Fixed it I think.

[the remark this answers is not in the document]
Comment (bcameron1231) [reply packages/sp/appcatalog/types.ts]: Tests well for me.

**Comment by bcameron1231** [comment]
You did check in the debug main.ts file, not sure if you meant to.

**Comment by bcameron1231** [comment]
The debug main.ts was committed, I assume that was an accident.

Everything else tests correctly for me!

**Pull request merged**